src/crimson/os/futurized_store.h

```diff
@@ -146,6 +147,9 @@ class FuturizedStore {
   virtual seastar::future<int, std::string> read_meta(const std::string& key) = 0;
   virtual uuid_d get_fsid() const  = 0;
   virtual unsigned get_max_attr_name_length() const = 0;
+  virtual ObjectStore* get_objectstore() {
```

Comment:
i am not sure we want to shape the interface this way.

Reply:
I agree with you, this is kind of a reluctant action. I added this method because PGLog::rebuild_missing_set_with_deletes() needs to call ObjectStore::getattr(), this call seems inevitable, because we have to determine the object's current on-disk version when rebuilding missing_set during the boot up phase, so I think we have to get the ObjectStore from FuturizedStore as long as we don't touch the current implementation of the peering. The only other way I can think of, for now, is that we do some total refactor of the peering state machine's reaction to AdvMap when it's in the Reset state and let the state machine allow other modules to pass it the versions of those objects, which seems to involve a large scale modification. Am I right about this? Thanks:-)

Reply:
@tchaikov I tried to add a new rebuild_missing_set_with_deletes method which uses FuturizedStore::get_attr(), but because this method maybe invoked from Reset::react(), so we have to call Future::get() method, which triggered an error during my test.

Reply:
> @tchaikov I tried to add a new rebuild_missing_set_with_deletes method which uses FuturizedStore::get_attr(), but because this method maybe invoked from Reset::react(), so we have to call Future::get() method, which triggered the assert "assert(thread);" in Future::do_wait() method during my test, I think this is because the Reset::react() method is not invoked from a seastar thread.

i see. you are talking about `PeeringState::Reset::react(const AdvMap& advmap)`, which calls `PeeringState::start_peering_interval()`. and in turn, `PeeringState::on_new_interval()` gets called. so `PG::rebuild_missing_set_with_deletes()` is called. this method checks the xattrs of objects in pg log for object info with older version.

but i think we can do `PeeringState::on_new_interval()` in background guarded with a gate.

@athanatos what do you think?